### PR TITLE
incremental restyle: Hoist most styling functionality from TNode to TElement

### PR DESCRIPTION
**Pull request by bholley**
This is a continuation of the work in #13951. I'm separating it out into a separate PR since the aforementioned patch has a green try run and this one doesn't yet.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13956)

### CODE REVIEW

**Comment by bholley** [comment]
r? @emilio 

(Review #13951 first)

File: components/layout/wrapper.rs [hunk 1]
```diff
@@ -43,49 +43,50 @@ pub type NonOpaqueStyleAndLayoutData = *mut AtomicRefCell<PersistentLayoutData>;
...
-        if self.borrow_layout_data().is_none() {
-            let ptr: NonOpaqueStyleAndLayoutData =
-                Box::into_raw(box AtomicRefCell::new(PersistentLayoutData::new()));
-            let opaque = OpaqueStyleAndLayoutData {
-                ptr: unsafe { NonZero::new(ptr as *mut AtomicRefCell<PartialPersistentLayoutData>) }
-            };
-            self.init_style_and_layout_data(opaque);
-        }
+impl<T: GetLayoutData> LayoutNodeHelpers for T {
+    fn ensure_layout_data(&self) -> NonOpaqueStyleAndLayoutData {
```
Comment (emilio) [inline components/layout/wrapper.rs]: So I'm slightly scared about this change in the sense that now calling `borrow_layout_data` can trigger a non thread-safe allocation, so I'll have to check this much more carefully.

File: components/layout_thread/lib.rs [hunk 2]
```diff
@@ -1111,8 +1111,8 @@ impl LayoutThread {
 
                         let mut current = node.parent_node();
                         while let Some(node) = current {
-                            if node.has_dirty_descendants() { break; }
-                            unsafe { node.set_dirty_descendants(); }
+                            if node.is_element() && node.as_element().unwrap().has_dirty_descendants() { break; }
```
Comment (emilio) [inline components/layout_thread/lib.rs]: Given we're checking on `parent_node`, this is guaranteed to be an element, right?

**Review by emilio — COMMENTED** [review]
The idea looks good to me, but I need to check with much more care before stamping this.

**Comment by emilio** [comment]
@bors-servo: retry

 * `bash: tests/wpt/run.sh: No such file or directory`?

**Comment by bholley** [comment]
> bash: tests/wpt/run.sh: No such file or directory

@bors-servo retry

File: components/layout/wrapper.rs (hunk 1 above)
Comment (bholley) [inline components/layout/wrapper.rs]: Great point. This observation escaped me in my post-midnight haze of getting this over the line. Let me take a crack at fixing this somehow.

**Comment by larsbergstrom** [comment]
@bors-servo retry
- Logs were busted, picked up https://github.com/servo/saltfs/pull/525

File: ports/geckolib/glue.rs [hunk 3]
```diff
@@ -253,7 +254,20 @@ pub extern "C" fn Servo_StyleSheet_Release(sheet: RawServoStyleSheetBorrowed) ->
...
      -> ServoComputedValuesStrong {
     let node = GeckoNode(node);
-    let data = node.borrow_data();
+
+    // Gecko erroneously calls this function from ServoRestyleManager::RecreateStyleContexts.
+    // We plan to fix that, but just support it for now until that code gets rewritten.
+    if node.is_text_node() {
+        error!("Don't call Servo_ComputedValue_Get() for text nodes");
+        let parent = node.parent_node().unwrap().as_element().unwrap();
+        let parent_cv = parent.borrow_data().map_or(Arc::new(ComputedValues::initial_values().clone()),
```
Comment (emilio) [inline ports/geckolib/glue.rs]: nit: `map_or_else` avoids creating the `Arc` unconditionally.

File: components/script/layout_wrapper.rs [hunk 4]
```diff
@@ -551,6 +550,18 @@ impl<'le> ServoLayoutElement<'le> {
...
         }
     }
+
+    fn mutate_data(&self) -> Option<AtomicRefMut<NodeData>> {
+        self.get_style_data().map(|d| d.borrow_mut())
+    }
+
+    fn get_partial_layout_data(&self) -> Option<&AtomicRefCell<PartialPersistentLayoutData>> {
+        unsafe {
+            self.get_style_and_layout_data().map(|d| {
```
Comment (emilio) [inline components/script/layout_wrapper.rs]: nit: Feel free to remove the braces here.

File: ports/geckolib/glue.rs [hunk 5]
```diff
@@ -106,24 +104,26 @@ fn restyle_subtree(node: GeckoNode, raw_data: RawServoStyleSetBorrowed) {
...
-        parallel::traverse_dom::<GeckoNode, RecalcStyleOnly>(node, &shared_style_context,
-                                                             per_doc_data.work_queue.as_mut().unwrap());
+        parallel::traverse_dom::<_, RecalcStyleOnly>(element.as_node(), &shared_style_context,
+                                                     per_doc_data.work_queue.as_mut().unwrap());
     }
 }
 
 #[no_mangle]
 pub extern "C" fn Servo_RestyleSubtree(node: RawGeckoNodeBorrowed,
                                        raw_data: RawServoStyleSetBorrowed) -> () {
```
Comment (emilio) [inline ports/geckolib/glue.rs]: I guess it'd be nice to pass an Element eventually.

File: ports/geckolib/glue.rs [hunk 6]
```diff
@@ -158,8 +158,9 @@ pub extern "C" fn Servo_StyleWorkerThreadCount() -> u32 {
 
 #[no_mangle]
 pub extern "C" fn Servo_Node_ClearNodeData(node: RawGeckoNodeBorrowed) -> () {
```
Comment (emilio) [inline ports/geckolib/glue.rs]: Same here, I guess doing both as a followup is fine.

File: components/style/gecko/wrapper.rs [hunk 7]
```diff
@@ -387,6 +297,55 @@ impl<'le> GeckoElement<'le> {
         let extra_data = ParserContextExtraData::default();
         parse_style_attribute(value, &base_url, Box::new(StdoutErrorReporter), extra_data)
     }
+
+    fn flags(&self) -> u32 {
+        self.raw_node()._base._base_1.mFlags
+    }
+
+    fn raw_node(&self) -> &RawGeckoNode {
+        &(self.0)._base._base._base
```
Comment (emilio) [inline components/style/gecko/wrapper.rs]: nit: Are these parentheses needed?

**Review by emilio — APPROVED** [review]
Looks good to me, r=me with nits addressed.

File: ports/geckolib/glue.rs (hunk 6 above)
Comment (bholley) [inline ports/geckolib/glue.rs]: Yep!

File: components/style/gecko/wrapper.rs (hunk 7 above)
Comment (bholley) [inline components/style/gecko/wrapper.rs]: Surprisingly yes - weird interaction of newtype and underscore-prefixed identifiers: https://is.gd/T1Tqct

File: ports/geckolib/glue.rs (hunk 3 above)
Comment (bholley) [inline ports/geckolib/glue.rs]: Fixed.

File: components/script/layout_wrapper.rs (hunk 4 above)
Comment (bholley) [inline components/script/layout_wrapper.rs]: The unsafe block right? Fixed.

**Comment by bholley** [comment]
@bors-servo r=emilio

Thanks for the review!

**Comment by bholley** [comment]
@bors-servo p=1

More refactorings on top.

**Comment by jdm** [comment]
```
error[E0133]: dereference of raw pointer requires unsafe function or block
   --> /home/servo/buildbot/slave/linux-dev/build/components/script/layout_wrapper.rs:560:14
    |
560 |             &**d.ptr
    |              ^^^^^^^ dereference of raw pointer

error: aborting due to previous error
```

File: components/script/layout_wrapper.rs (hunk 4 above)
Comment (emilio) [inline components/script/layout_wrapper.rs]: Nope the unsafe block is needed because of the deref, I meant (`map(|d| &**d.ptr)`)

**Comment by bholley** [comment]
Whoops - I forget it was a raw pointer and not Deref impls.

@bors-servo r=emilio

**Comment by emilio** [comment]
@bors-servo: retry 

 * #13360 + an IO error.

**Pull request merged**